src/io/category_encoding_provider.cpp

```diff
@@ -0,0 +1,1330 @@
+          Log::Fatal("categorical_feature is not a number,\n"
+                    "if you want to use a column name,\n"
+                    "please add the prefix \"name:\" to the column name");
+        }
+        categorical_features.emplace(tmp);
+      }
+    }
+  }
+
+  for (const int fid : ignore_features) {
+    if (categorical_features.count(fid)) {
+      categorical_features.erase(fid);
+    }
+  }
+
+  // reset categorical features for Config
+  if (config->categorical_feature.size() > 0) {
+    if (!keep_raw_cat_method_) {
+      config->categorical_feature.clear();
+      config->categorical_feature.shrink_to_fit();
+    }
+  }
+
+  categorical_features_.clear();
+  for (const int fid : categorical_features) {
+    if (fid < num_original_features_) {
+      categorical_features_.push_back(fid);
+    } else {
+      Log::Warning("Categorical feature index %d is no less than the actual feature number %d, and will be ignored.",
+        fid, num_original_features_);
+    }
+  }
+  std::sort(categorical_features_.begin(), categorical_features_.end());
+  is_categorical_feature_.clear();
+
+  return label_idx;
+}
+
+void CategoryEncodingProvider::PrepareCategoryEncodingStatVectors() {
+  is_categorical_feature_.clear();
+  is_categorical_feature_.resize(num_original_features_, false);
+  for (const int fid : categorical_features_) {
+    if (fid < num_original_features_) {
+      is_categorical_feature_[fid] = true;
+    }
+  }
+  fold_prior_.resize(config_.num_target_encoding_folds + 1, 0.0f);
+  if (category_encoders_.size() > 0) {
+    // prepare to accumulate target encoding statistics
+    fold_label_sum_.resize(config_.num_target_encoding_folds + 1, 0.0f);
+    fold_num_data_.resize(config_.num_target_encoding_folds + 1, 0);
+    if (!accumulated_from_file_) {
+      thread_fold_label_sum_.resize(num_threads_);
+      thread_fold_num_data_.resize(num_threads_);
+      thread_count_info_.resize(num_threads_);
+      thread_label_info_.resize(num_threads_);
+      for (int thread_id = 0; thread_id < num_threads_; ++thread_id) {
+        thread_fold_label_sum_[thread_id].resize(config_.num_target_encoding_folds + 1, 0.0f);
+        thread_fold_num_data_[thread_id].resize(config_.num_target_encoding_folds + 1, 0.0f);
+      }
+    }
+    for (const int fid : categorical_features_) {
+      if (fid < num_original_features_) {
+        count_info_[fid].resize(config_.num_target_encoding_folds + 1);
+        label_info_[fid].resize(config_.num_target_encoding_folds + 1);
+        if (!accumulated_from_file_) {
+          for (int thread_id = 0; thread_id < num_threads_; ++thread_id) {
+            thread_count_info_[thread_id][fid].resize(config_.num_target_encoding_folds + 1);
+            thread_label_info_[thread_id][fid].resize(config_.num_target_encoding_folds + 1);
+          }
+        }
+      }
+    }
+  }
+}
+
+void CategoryEncodingProvider::SyncEncodingStat(std::vector<std::unordered_map<int, double>>* fold_label_sum_ptr,
```

Comment:
Not quite understand this SyncEncodingStat method. Not sure we can add more comments for it?

Reply:
Resoved via 3e91f68